docs/static/field-reference.asciidoc

```diff
@@ -0,0 +1,129 @@
+
+_Field References_ can be expressed literally within <<configuration#conditionals,_Conditional_>> statements in your pipeline configurations,
+as string arguments to your pipeline plugins, or within sprintf statements that will be used by your pipeline plugins:
+
+[source,pipelineconf]
+filter {
+  #  +----literal----+     +----literal----+
+  #  |               |     |               |
+  if [@metadata][date] and [@metadata][time] {
+    mutate {
+      add_field {
+        "[@metadata][timestamp]" => "%{[@metadata][date]} %{[@metadata][time]}"
+      # |                      |    |  |               |    |               | |
+      # +----string-argument---+    |  +--field-ref----+    +--field-ref----+ |
+      #                             +-------- sprintf format string ----------+
+      }
+    }
+  }
+}
+
+[[formal-grammar]]
+=== Formal Grammar
+
+Below is the formal grammar of the Field Reference, with notes and examples.
+
+[[formal-grammar-field-reference-literal]]
+==== Field Reference Literal
+
+A _Field Reference Literal_ is a sequence of one or more _Path Fragments_ that can be used directly in Logstash pipeline <<configuration#conditionals,conditionals>> without any additional quoting (e.g. `[request]`, `[response][status]`).
+
+[source,antlr]
+fieldReferenceLiteral
+  : ( pathFragment )+
+  ;
+
+[[formal-grammar-field-reference]]
+==== Field Reference (Event APIs)
+
+The Event API's methods for manipulating the fields of an event or using the sprintf syntax are more flexible than the pipeline grammar in what they accept as a Field Reference.
+Top-level fields can be referenced directly by their _Field Name_ without the square brackets, and there is some support for _Composite Field References_, simplifying use of programmatically-generated Field References.
+
+A _Field Reference_ for use with the Event API is therefore one of:
+
+ - a single _Field Reference Literal_; OR
+ - a single _Field Name_ (referencing a top-level field); OR
+ - a single _Composite Field Reference_.
+
+[source,antlr]
+eventApiFieldReference
+  : fieldReferenceLiteral
+  | fieldName
+  | compositeFieldReference
+  ;
+
+[[formal-grammar-path-fragment]]
+==== Path Fragment
+
+A _Path Fragment_ is a _Field Name_ wrapped in square brackets (e.g., `[request]`).
+
+[source,antlr]
+pathFragment
+  : '[' fieldName ']'
+  ;
+
+[[formal-grammar-field-name]]
+==== Field Name
+
+A _Field Name_ is a sequence of characters that are _not_ square brackets (`[` or `]`).
+
+[source,antlr]
+fieldName
+  : ( ~( '[' | ']' ) )+
+  ;
+
+[[formal-grammar-event-api-composite-field-reference]]
+==== Composite Field Reference
+
+In some cases, it may be necessary to programmatically _compose_ a Field Reference from one or more Field References,
+such as when manipulating fields in a plugin or while using the Ruby Filter plugin and the Event API.
+
+[source,ruby,indent=4]
+    fieldReference = "[path][to][deep nested field]"
```

Comment:
WRT `indent=4` .   It appears that asciidoc does not support this notation.
When I get a successful build, we can see what it looks like and find a way to achieve the result you want using a different approach.